File: Python/matlab/clusterandregress.m
clear; clc; close all;

is_qian_bei = 1
% 读取数据
if is_qian_bei == 1
    [data1, name1] = read_qian_bei(ok); % 3#
else
    [data1, name1] = read_gao_jia(ok); % (+
end

% 聚类分析
D = pdist(data1, 'minkowski', 2);
Z = linkage(D, 'ward');
H = dendrogram(Z, 'labels', name1);
set(H, 'color', 'k', 'linewidth', 1.3);

if is_qian_bei == 1
    title('3#&,.5?7', 'FontSize', 16);
else
    title('(+&,.5?7', 'FontSize', 16);
end

xlabel('*:$'', 'FontSize', 12);
ylabel('Scale', 'FontSize', 12);
k = 2;
T = cluster(Z, k);

for i = 1:k
    tm = find(T == i);
    tm = tm';
    fprintf(2, '&%d.&<x\n', i);
    for j = 1:length(tm)
        fprintf("%s ", name1{tm(j)});
    end
    fprintf("\n");
end

% 进一步的聚类分析
z1 = find(T == 1);
D = pdist(data1(z1, :), 'minkowski', 2);
Z = linkage(D, 'ward');
name11 = name1(z1);
H = dendrogram(Z, 'labels', name11);
xlabel('*:$');
ylabel('Scale');
C = cophenet(Z, D);
k = 2;
T1 = cluster(Z, k);

for i = 1:k
    tm = find(T1 == i);
    tm = tm';
    fprintf('&%d.&<x\n', i);
    for j = 1:length(tm)
        fprintf("%s ", name11{tm(j)});
    end
    fprintf("\n");
end

z2 = find(T == 2);
D = pdist(data1(z2, :), 'minkowski', 2);
Z = linkage(D, 'ward');
name12 = name1(z2);
H = dendrogram(Z, 'labels', name12);
xlabel('*:$');
ylabel('Scale');
C = cophenet(Z, D);
k = 2;
T2 = cluster(Z, k);
T2 = T2 + 2;

for i = 3:k+2
    tm = find(T2 == i);
    tm = tm';
    fprintf('&%d.&<x\n', i);
    for j = 1:length(tm)
        fprintf("%s ", name12{tm(j)});
    end
    fprintf("\n");
end

z12 = [z1; z2];
T12 = [T1; T2];
[~, ind] = sort(z12);
T12 = T12(ind);

% 提取特征均值
ok = 0;
if is_qian_bei == 1
    [data1, name1] = read_qian_bei(ok); % 3#
else
    [data1, name1] = read_gao_jia(ok); % (+
end

fea = ones(14, 4);
for i = 1:14
    for t = 1:4
        fea(i, t) = mean(data1(T12 == t, i));
    end
end

fea = fea';
fea = [fea(1, :); fea(2, :); fea(4, :); fea(3, :)];

% 拟合多项式并绘图
x0 = 1:4;
x1 = 1:0.2:4;
fea_name = {
    "' ;!(SiO2)"
    ";!*(Na2O)"
    ";!*(K2O)"
    ";!*(CaO)"
    ";!*(MgO)"
    ";!*(Al2O3)"
    ";!*(Fe2O3)"
    ";!*(CuO)"
    ";!*(PbO)"
    ";!*(BaO)"
    "8×;!*(P2O5)"
    ";!*(SrO)"
    ";!*(SnO2)"
    "' ;!(SO2)"
};

a = 1; b = 14;
p = zeros(14, 3);
for i = a:b
    p(i, :) = polyfit(x0, fea(:, i), 2);
    y1 = polyval(p(i, :), x1);
    figure; 
    H = plot(x0, fea(:, i), '*', 'LineWidth', 5); 
    hold on; 
    plot(x1, y1, 'o');
    legend('实际值', '拟合值');
    xlabel('*:$');
    title(fea_name{i});
end

% 读取预测数据并处理
[qian_bei3, qian_bei4, gao_jia3, gao_jia4] = read_pred();
qian_bei_pred4 = zeros(size(qian_bei4));
gao_jia_pred4 = zeros(size(gao_jia4));
qian_bei_pred3 = zeros(size(qian_bei3));
gao_jia_pred3 = zeros(size(gao_jia3));

if is_qian_bei == 1
    % MgO K2O
    for i = 1:size(qian_bei4, 1)
        for j = a:b
            if j == 5 || j == 3
                qian_bei_pred4(i, j) = qian_bei4(i, j);
                continue;
            end
            if qian_bei4(i, j) > 0
                qian_bei_pred4(i, j) = power(qian_bei4(i, j) - getY(p(j, :), 4), 1/2) + getY(p(j, :), 1);
            else
                qian_bei_pred4(i, j) = -power(-qian_bei4(i, j) - getY(p(j, :), 4), 1/2) + getY(p(j, :), 1);
            end
        end
    end

    for i = 1:size(qian_bei3, 1)
        for j = a:b
            if j == 5 || j == 3
                qian_bei_pred3(i, j) = qian_bei3(i, j);
                continue;
            end
            if qian_bei3(i, j) > 0
                qian_bei_pred3(i, j) = power(qian_bei3(i, j) - getY(p(j, :), 4), 1/2) + getY(p(j, :), 1);
            else
                qian_bei_pred3(i, j) = -power(-qian_bei3(i, j) - getY(p(j, :), 4), 1/2) + getY(p(j, :), 1);
            end
        end
    end
    writematrix(qian_bei_pred4, '../qian_bei4.xlsx')
    writematrix(qian_bei_pred3, '../qian_bei3.xlsx')
else
    % SO2 SnO2 Na2O
    for i = 1:size(gao_jia4, 1)
        for j = a:b
            if j == 14 || j == 13 || j == 2
                gao_jia_pred4(i, j) = gao_jia4(i, j);
                continue;
            end
            if gao_jia4(i, j) > 0
                gao_jia_pred4(i, j) = power(gao_jia4(i, j) - getY(p(j, :), 4), 1/2) + getY(p(j, :), 1);
            else
                gao_jia_pred4(i, j) = -power(-gao_jia4(i, j) - getY(p(j, :), 4), 1/2) + getY(p(j, :), 1);
            end
        end
    end

    for i = 1:size(gao_jia3, 1)
        for j = a:b
            if j == 14 || j == 13 || j == 2
                gao_jia_pred3(i, j) = gao_jia3(i, j);
                continue;
            end
            if gao_jia3(i, j) > 0
                gao_jia_pred3(i, j) = power(gao_jia3(i, j) - getY(p(j, :), 4), 1/2) + getY(p(j, :), 1);
            else
                gao_jia_pred3(i, j) = -power(-gao_jia3(i, j) - getY(p(j, :), 4), 1/2) + getY(p(j, :), 1);
            end
        end
    end
    writematrix(gao_jia_pred4, '../gao_jia_pred4.xlsx')
    writematrix(gao_jia_pred3, '../gao_jia_pred3.xlsx')
end

function y = getY(p, x)
    y = polyval(p, x);
end

File: Python/matlab/clusterandregress.py
import numpy as np
from scipy.spatial.distance import pdist, squareform
from scipy.cluster.hierarchy import linkage, dendrogram, fcluster, cophenet
import matplotlib.pyplot as plt

# 定义一些自定义函数
def read_qian_bei(ok):
    # 这里需要实现读取千贝数据的函数
    pass

def read_gao_jia(ok):
    # 这里需要实现读取高甲数据的函数
    pass

def read_pred():
    # 这里需要实现读取预测数据的函数
    pass

def getY(p, x):
    # 计算多项式的值
    return np.polyval(p, x)

# 主程序
is_qian_bei = 0
ok = 1

if is_qian_bei == 1:
    data1, name1 = read_qian_bei(ok)
else:
    data1, name1 = read_gao_jia(ok)

D = pdist(data1, 'minkowski', 2)
Z = linkage(D, 'ward')
H = dendrogram(Z, labels=name1)
plt.setp(H, color='k', linewidth=1.3)

if is_qian_bei == 1:
    plt.title('千贝分类', fontsize=16)
else:
    plt.title('高甲分类', fontsize=16)

plt.xlabel('City', fontsize=12)
plt.ylabel('Scale', fontsize=12)
plt.show()

k = 2
T = fcluster(Z, k, criterion='maxclust')

for i in range(1, k + 1):
    tm = np.where(T == i)[0]
    print(f'Cluster {i}:', ', '.join(name1[j] for j in tm))

# 对第一个簇进行重新聚类
z1 = np.where(T == 1)[0]
D = pdist(data1[z1, :], 'minkowski', 2)
Z = linkage(D, 'ward')
name11 = [name1[j] for j in z1]
H = dendrogram(Z, labels=name11)
plt.xlabel('City')
plt.ylabel('Scale')
plt.show()

T1 = fcluster(Z, k, criterion='maxclust')

for i in range(1, k + 1):
    tm = np.where(T1 == i)[0]
    print(f'Cluster {i}:', ', '.join(name11[j] for j in tm))

# 对第二个簇进行重新聚类
z2 = np.where(T == 2)[0]
D = pdist(data1[z2, :], 'minkowski', 2)
Z = linkage(D, 'ward')
name12 = [name1[j] for j in z2]
H = dendrogram(Z, labels=name12)
plt.xlabel('City')
plt.ylabel('Scale')
plt.show()

T2 = fcluster(Z, k, criterion='maxclust')
T2 += 2

for i in range(3, k + 3):
    tm = np.where(T2 == i)[0]
    print(f'Cluster {i}:', ', '.join(name12[j] for j in tm))

z12 = np.concatenate([z1, z2])
T12 = np.concatenate([T1, T2])
ind = np.argsort(z12)
T12 = T12[ind]

# 计算每个簇的均值
fea = np.ones((14, 4))
for i in range(14):
    for t in range(4):
        fea[i, t] = np.mean(data1[T12 == t, i])

fea = fea.T
fea = fea[[0, 1, 3, 2], :]

x0 = np.arange(1, 5)
x1 = np.arange(1, 4.2, 0.2)

fea_name = [
    "SiO2", "Na2O", "K2O", "CaO", "MgO", "Al2O3", "Fe2O3", "CuO", 
    "PbO", "BaO", "P2O5", "SrO", "SnO2", "SO2"
]

a, b = 1, 14
p = np.zeros((14, 3))

for i in range(a, b + 1):
    p[i - 1, :] = np.polyfit(x0, fea[:, i - 1], 2)
    y1 = np.polyval(p[i - 1, :], x1)
    plt.figure()
    plt.plot(x0, fea[:, i - 1], '*', linewidth=5)
    plt.plot(x1, y1, 'o')
    plt.legend(['Original', 'Fitted'])
    plt.xlabel('Cluster')
    plt.title(fea_name[i - 1])
    plt.show()

qian_bei3, qian_bei4, gao_jia3, gao_jia4 = read_pred()
qian_bei_pred4 = np.zeros_like(qian_bei4)
gao_jia_pred4 = np.zeros_like(gao_jia4)
qian_bei_pred3 = np.zeros_like(qian_bei3)
gao_jia_pred3 = np.zeros_like(gao_jia3)

if is_qian_bei == 1:
    for i in range(qian_bei4.shape[0]):
        for j in range(a, b + 1):
            if j == 5 or j == 3:
                qian_bei_pred4[i, j] = qian_bei4[i, j]
                continue
            if qian_bei4[i, j] > 0:
                qian_bei_pred4[i, j] = np.sqrt(qian_bei4[i, j] - getY(p[j - 1, :], 4)) + getY(p[j - 1, :], 1)
            else:
                qian_bei_pred4[i, j] = -np.sqrt(-qian_bei4[i, j] - getY(p[j - 1, :], 4)) + getY(p[j - 1, :], 1)
    
    for i in range(qian_bei3.shape[0]):
        for j in range(a, b + 1):
            if j == 5 or j == 3:
                qian_bei_pred3[i, j] = qian_bei3[i, j]
                continue
            if qian_bei3[i, j] > 0:
                qian_bei_pred3[i, j] = np.sqrt(qian_bei3[i, j] - getY(p[j - 1, :], 4)) + getY(p[j - 1, :], 1)
            else:
                qian_bei_pred3[i, j] = -np.sqrt(-qian_bei3[i, j] - getY(p[j - 1, :], 4)) + getY(p[j - 1, :], 1)
    
    np.savetxt('../qian_bei4.csv', qian_bei_pred4, delimiter=',')
    np.savetxt('../qian_bei3.csv', qian_bei_pred3, delimiter=',')
else:
    for i in range(gao_jia4.shape[0]):
        for j in range(a, b + 1):
            if j == 14 or j == 13 or j == 2:
                gao_jia_pred4[i, j] = gao_jia4[i, j]
                continue
            if gao_jia4[i, j] > 0:
                gao_jia_pred4[i, j] = np.sqrt(gao_jia4[i, j] - getY(p[j - 1, :], 4)) + getY(p[j - 1, :], 1)
            else:
                gao_jia_pred4[i, j] = -np.sqrt(-gao_jia4[i, j] - getY(p[j - 1, :], 4)) + getY(p[j - 1, :], 1)
    
    for i in range(gao_jia3.shape[0]):
        for j in range(a, b + 1):
            if j == 14 or j == 13 or j == 2:
                gao_jia_pred3[i, j] = gao_jia3[i, j]
                continue
            if gao_jia3[i, j] > 0:
                gao_jia_pred3[i, j] = np.sqrt(gao_jia3[i, j] - getY(p[j - 1, :], 4)) + getY(p[j - 1, :], 1)
            else:
                gao_jia_pred3[i, j] = -np.sqrt(-gao_jia3[i, j] - getY(p[j - 1, :], 4)) + getY(p[j - 1, :], 1)
    
    np.savetxt('../gao_jia_pred4.csv', gao_jia_pred4, delimiter=',')
    np.savetxt('../gao_jia_pred3.csv', gao_jia_pred3, delimiter=',')
